### DISCUSSION TITLE
Dryioc and Microsoft Dependency injection with servicekey

### DISCUSSION THREAD

**Discussion by jfield-igniteous** [dadhi/DryIoc#626] [Q&A]
I have a Dryioc container into which I register a number of components. One of these registrations uses a servicekey lookup to resolve a dependency at runtime.

 builder.Register<IErpData, ErpData>(Reuse.Singleton, serviceKey: "local");
 var erpAccessParams = Parameters.Of
                .Name("erpData", (request) =>
                {
                    // Perform late lookup to determine which concrete instance to inject
                    var erpAccess = builder.Resolve<IErpData>(serviceKey:"local");
                    return erpAccess;
                });

builder.Register<QueueActor>(made: Made.Of(parameters: erpAccessParams));

A MS IHostBuilder instance is used to set up the initial container definition for MS registrations via a call to IHostBuilder.UseServiceProviderFactory(new DryIocServiceProviderFactory()); 

The odd behaviour I'm seeing is that the Parameters.Of.Name lambda is never called to revolve the constructor argument in some situations.

If I use the following:

 builder.RegisterDelegate<IErpData>((context) =>
  {
      return context.Resolve<ErpData>(); 
  });    
builder.Register<QueueActor>();

Then the resolution works. It seems to be an issue on matching parameters. How does this constructor parameter selection process work in combination with MS DI container?

**Comment by dadhi** [comment]
@jfield-igniteous 

I would avoid this kind of context based resolutions:
```
builder.Register<IErpData, ErpData>(Reuse.Singleton, serviceKey: "local");
var erpAccessParams = Parameters.Of
.Name("erpData", (request) =>
{
// Perform late lookup to determine which concrete instance to inject
var erpAccess = builder.Resolve(serviceKey:"local");
return erpAccess;
});

builder.Register(made: Made.Of(parameters: erpAccessParams));
```

Here, you are actually resolving the parameter value of a specific key as a resolution root and then using it for injection.

Instead, it is much simpler to configure the injection of parameter `erpData` directly.
```cs
builder.Register<Foo>(made: Parameters.Of.Name("erpData", serviceKey: "local"));
```

If you need the Runtime based condition (**do you really?**), then add 2 registrations and selective condition:
```cs
builder.Register<Foo>(made: Parameters.Of.Name("erpData", serviceKey: "local"), setup: Setup.With(condition: _ => env.IsDev));
builder.Register<Foo>(made: Parameters.Of.Name("erpData", serviceKey: "prod"), setup: Setup.With(condition: _ => env.IsProd));
```

**Reply by jfield-igniteous** [reply]
Thank you for that feedback. I will adjust how I make these types of registrations . The resolution is based on configuration data that is read in at runtime to determine which of the data providers to use. E.g Sql or Access for example.

However the issue I am seeing is that the parameter selectors are not be executed at all when creating the components via a MS dependency injection container (via an external framework). I have a suspicion that the way the framework code is passing in an empty array of parameters for the constructor arguments is steering ctor resolution to a default ctor (without the parameter)

Is there a way to steer this resolution by ignoring the parameters being passed in from the framework? Or to arrange the priority of containers definitions over those passed in? If so, I'll try and see if this steers the ctor matcher correctly.

**Comment by dadhi** [comment]
The support for keyed services for MS.DI is not yet released. But it only matters if the client calls your code expecting a keyed service. If you register via DryIoc container and not via the ServiceCollection  - it should work.

So to undertstand your particular issue I need more context.

**Reply by jfield-igniteous** [reply]
I'm using Akka.NET and am having to migrate my code to use their dependency injection model - that relies on the MS DI model. I am forced to use MS Hosting libraries - and many of the registrations are defined within these MS libraries. 

I've been burned too many times over the years by MS changing priorities and deprecating code. So I perform all the application component registrations using Dryioc. I have to merge these registrations into the MS DI container view. No biggie in itself.

However I've just noticed that with the migration, when Itry and create an actor , any constructor arguments that require a parameter selector are not being matched/fired, and it seems to be using the default constructor. A line of investigation that I am following here is whether the selection of the constructor to use is being skewed towards the default constructor rather than the definition defined in the container.

If I donlt use parameter selections, e.g. use a delegate, then the resolution works ok. But this approach will only work for typed parameters and not named parameters (where the value to inject in is just a string).

Anther option may be to move towards an IOption<> pattern for injecting in constructor arguments to get around this issue. It just seems odd that in this case the named parameter selector is being ignored.

**Reply by dadhi** [reply]
What is the container configuration? You may do `container.ToString()` to see.
I am interested, does it have the constructor selection rule.

If you are integrating with MS.DI hosting... how are you doing it?
Also, MS.DI integration may produce the new container with the adjusted rules (conforming to MS.DI conventions) and then you may operate on the wrong container instance.

**Reply by jfield-igniteous** [reply]
This is how I configure the container. I hook in a class instance that gets called to register the components. Thats the bootstrapper.

![Screenshot 2024-01-22 at 18 41 00](https://github.com/dadhi/DryIoc/assets/157137946/aaafefec-c6ac-414a-83a7-f65ff2ed7e3d)

**Reply by jfield-igniteous** [reply]
![Screenshot 2024-01-22 at 18 43 26](https://github.com/dadhi/DryIoc/assets/157137946/6dfd27bf-2499-4011-8900-06e76d9ee5dc)

Its not pretty but here is the hooking in the Actor dependency resolver
![Screenshot 2024-01-22 at 18 43 38](https://github.com/dadhi/DryIoc/assets/157137946/1da5e42f-95c3-4edf-b44c-90b1d2444845)

**Reply by jfield-igniteous** [reply]
The bootstrapper allows me to centralise my DI configuration and to also support dynamically loading registrations at runtime from registration plugins

**Reply by dadhi** [reply]
@jfield-igniteous 

> Its not pretty but here is the hooking in the Actor dependency resolver

Here the `IContainer builder` may be the different instance from the `adaptedContainer`.

**Reply by jfield-igniteous** [reply]
Yes I could imagine that. I only added that line to see if it made any difference to the issues with resolutions on the parameters for ctors. It wasn't in the original version.

What I did do as a test was to just create the DependencyResolver manually (which wraps the Dryioc container) to see if the parameter selection works. Which it does as long as you don't use the Props<>{) operation.And this is where I am seeing differences in behaviour.

Hence my earlier point around the selection of constructor to use when activating the type definition. It seems that whatever this method is doing, is affecting the constructor selection

**Reply by jfield-igniteous** [reply]
How does DryIoc determine which constructor that best matches? i.e. What data would be required to skew the constructor selection to use the default, when another (non-default) constructor existed?

**Reply by dadhi** [reply]
Ms.DI will add the rule to select most resolvable constructor. 
By default DryIoc throws a sensible exception if faced with multiple constructors.
That's why it makes sense to introspect the actual container and its rules.

**Reply by jfield-igniteous** [reply]
container without scope
 with Rules with {TrackingDisposableTransients, SelectLastRegisteredFactory} and without {ThrowOnRegisteringDisposableTransient, VariantGenericTypesInResolvedCollection}
 with FactorySelector=SelectLastRegisteredFactory
 with Made={FactoryMethod=ConstructorWithResolvableArguments}

**Reply by dadhi** [reply]
Also, what is the visibility of ctors?
If default is public and the rest is not, DryIoc will select the default.

**Reply by dadhi** [reply]
Ok...
FactoryMethod=ConstructorWithResolvableArguments

will select the public over non public as well. Then you need a variation of the rule with IncludingNonPublic at the end.

**Reply by jfield-igniteous** [reply]
The constructor I am defining is public, the default is protected (which is on the base class which is declared as abstract)

**Reply by jfield-igniteous** [reply]
![Uploading Screenshot 2024-01-22 at 19.24.15.png…]()

**Reply by jfield-igniteous** [reply]
Its almost as if the information the Akka.Net framework is passing is suggesting to DryIoc the selection of the default. But only if there is a Parameter selection rule defined in DryIoc.

**Reply by dadhi** [reply]
ok. 
Do a simple test. 
Register a sample keyed service in adaptedContainer after you got it. 
Register a sample consumer service with the Parameters.Of.Name blah blah.
Then immediately resolve the consumer.
Assert what you'd get.

**Reply by jfield-igniteous** [reply]
One of the tests I have already done was to determine if the DependencyResvoler they created was causing the issue. Initially I created a component definition in DryIoc with a Parameters.Of.Name and did a resolve. That worked and created a simple POCO class instance. 

Their DependencyResolver accepts a IServiceProvider instance (which I created from the Container). There are two method types on this class: A GetService<T> which delegates down to the underlying IServiceProvider instance. This works fine.

And Props<T>() method which is needed to create actors. This is the operation that is failing.What I don't quite understand yet is how DryIoc selects a constructor if there is more than one available.

**Reply by jfield-igniteous** [reply]
How are Parameter selectors chosen to fire? If the same code in Akka.NET behaves differently based on the absence or presence of a parameter selector in DryIoc, via MS DI, that would suggest that the logic for picking a constructor is being influenced.

**Reply by dadhi** [reply]
With the rule in place, it selects the constructor with most parameters first where all parameters are resolved. So it tries to resolve the expressions for the constructor params first without creating the actual values for them. If some param resolution is failing, it will proceed to the next ctor with less params and ultimately to the default one (if present).

**Reply by jfield-igniteous** [reply]
If there are no parameter selectors on a class (but has a ctor argument that can be resolved via a registration in the container) this uses a different constructor resolver path? i.e. The parameter selection path happens before the standard ctor resolver?

**Reply by dadhi** [reply]
1st, What version of DryIoc are you using?

2nd,
> If there are no parameter selectors on a class (but has a ctor argument that can be resolved via a registration in the container) this uses a different constructor resolver path? i.e. The parameter selection path happens before the standard ctor resolver?

Did not get it. Please provide the example.

3rd, From the code alone, I see that most resolvable constructor selection takes the ParameterSelector into account. So it expects for parameter selector to at least provide something resolvable.

**Reply by jfield-igniteous** [reply]
I'm using 5.4.3 DryIoc.dll nuget.

I'm currently step debugging through Akka.Net, and down into Dryioc. From what I can see, the type I am attempting to resolve does exist in the underlying registry, keyed by its interface. I stepped through the register and could see that it created a keyed service entry.

In the container.ResolveAndCache() method it creates a request.CreateResolutionRoot for the servicetype that has no value for servicekey. The ResolveFactory() call calls down to GetServiceFactoryOrDefault(). The GetRuleSelectedServiceFactoryOrDefault() returns an entry that contains the factories entry for the given type, along with the key = "local" (where local was the previously registered service key for that type). 

At line 1184 factories.Length == 1 and therefore it attempts to match the given factory based on the request criteria passed in from above. However this does not match as the factory is keyed based on "local" and the serviceDetails variable passed in from above has null for ServiceKey.

So it appears that it isn't matching the previously registered type, because the servicekey passed from above is null. 


NOTE >>
This call chain above is triggerred by the ActivatorUtilities.CreateInstance(). It then extracts the parameter types associated with the type to create. It creates a ConstructorMatcher which then attempts to match the correct constructor. As part of this process it calls GetService() line 524 (ActivatorUtilities) with the parameter type.

![Screenshot 2024-01-23 at 11 07 45](https://github.com/dadhi/DryIoc/assets/157137946/8599853b-4088-4f7b-a9f0-f702c5210c33)

![Screenshot 2024-01-23 at 11 11 08](https://github.com/dadhi/DryIoc/assets/157137946/ed20dd67-4b54-4c12-9d7f-1966da6c2304)

![Screenshot 2024-01-23 at 11 13 26](https://github.com/dadhi/DryIoc/assets/157137946/ccd638c0-cc15-4044-9943-2421bd31490e)

![Screenshot 2024-01-23 at 11 14 18](https://github.com/dadhi/DryIoc/assets/157137946/22edf125-c2ef-4bd3-b93b-db1df6c40f84)


At this point Dryioc gets involved to create the parametertype, which in our case we have defined in Dryioc as a keyedservice. However the call from MS DI ActivatorUtilities.GetService is not passing across any context to DryIoc to tell it that it has any servicekey name associated with this. Hence why it fails to resolve.
![Screenshot 2024-01-23 at 11 14 34](https://github.com/dadhi/DryIoc/assets/157137946/92c16ea9-2eeb-440e-bdfa-a979f7ad443a)


Interestingly underneath the same block of code is a branch condition to do a GetKeyedService.
![Screenshot 2024-01-23 at 11 22 09](https://github.com/dadhi/DryIoc/assets/157137946/6b7922ab-e6b2-4353-a543-221a72c20512)

**Reply by dadhi** [reply]
Hi. So if your investigation found the actual bug, then it should be reproducible with the simple pair of service and dependency.
Could you please write a test, at least here, so that I can check it on my side?

**Reply by jfield-igniteous** [reply]
One approach to correct this is to steer MS DI to pass through the context to Dryioc, we would have to now pollute our application code with MS DI attributes (which has no MS DI dependencies at all up until now) to get it to play nicely with DryIoc (when it delegates down to Dryioc to fulfil the request). And Dryioc would have to implement keyed services as well (as a service provider).

Does this make sense?

**Reply by jfield-igniteous** [reply]
Another approach should be that Dryioc would have to add a new MS DI rule that enabled the GetService() call for the parameter type dependency to ignore any keyed registrations in order to resolve the dependency of parameters first. 

On attempting to resolve the type that implements the ctor your parameter selector logic would fire as today and resolve correctly? If there was no matching keyed component registration (as part of the normal activation of a Dryioc managed type) then the component instantiation would fail because it could not find the required keyed dependency

**Reply by jfield-igniteous** [reply]
The underlying issue here seems to me that MS DI ActivatorUtilities is orchestrating the individual parameter type resolution on the constructor via a loop, rather than delegating the type to be created to the serviceprovider (Dryioc) to resolve all of the dependencies in one go.

It assumes that it has the responsibility to resolve dependencies - piece by piece, because it is acting as an abstraction layer. rather than delegating the responsibility to the underlying container to resolve.

**Reply by dadhi** [reply]
@jfield-igniteous Next DryIoc version (v6.0.0) will support the `IKeyedServiceProvider` and its friends.
I would be interested to check how it works in your case. The version is close to be done, all tests are green - but I need to do the internal cleanups. 

I may release right now the preview version of V6 and the corresponding preview version of Dryioc.Microsoft.DependencyInjection v8.0.0-preview-01, so you would be able to check.

**Reply by jfield-igniteous** [reply]
There might be an interception point within Akka.NET to reroute the request to create an actor away from the MS DI ActivatorUtulities, and direct to the ServiceProvider instance (Dryioc) in this case. I will try this today.

I would prefer to avoid tightly coupling my code with MS DI just to make it work with Dryioc.

**Reply by jfield-igniteous** [reply]
As an aside, I really appreciate the fantastic work you've done (and continue to do) with the DryIoc product. You're always very responsive and helpful. Thank you.

**Reply by dadhi** [reply]
Anyway, I have pushed DryIoc v6.0.0-preview-06 and DryIoc.Microsoft.DependencyInjection v8.0.0-preview-01 to NuGet.

**Reply by jfield-igniteous** [reply]
I'm working on a basic POC for Akka.NET that wraps DryIoc directly to perform resolutions, without using ServiceProvider wrapper. It seems to work so far with the original configuration (parameter section and servicekeys). I'll keep you posted if I find any issues.

**Reply by dadhi** [reply]
Great, if you will publish a repo please send me a link.
